File: packages/client/src/stlink.ts
/** STMicroelectronics VID. ST-Link V2 and V2-1/V3 share it. */
const STLINK_VID = 0x0483;

/**
 * The shipped worker wasm still reports an unclaimed debug interface as
 * `open-failed`. Map that to `stlink-interface` so the dock does not reuse the
 * CMSIS-DAP v1 (HID) copy. A rebuilt worker that already sets the kind is unchanged.
 */
export function classifyStlinkInterfaceError(vendorId: number, error: unknown): unknown {
  if (!stlinkInterfaceFailure(vendorId, errorText(error))) return error;
  if (error instanceof Error) {
    (error as Error & { kind?: string }).kind = 'stlink-interface';
    return error;
  }
  return Object.assign(new Error(errorText(error)), { kind: 'stlink-interface' });
}

function stlinkInterfaceFailure(vendorId: number, message: string): boolean {
  if (vendorId !== STLINK_VID) return false;
  const text = message.toLowerCase();
  return (
    text.includes('endpoint not found') ||
    text.includes('endpointnotfound') ||
    text.includes('interface not found') ||
    text.includes('claiminterface') ||
    text.includes('could not be claimed')
  );
}

function errorText(error: unknown): string {
  if (error instanceof Error) return error.message;
  return String(error);
}

File: crates/probe-web-core/src/stlink.rs
//! Classify an ST-Link open failure that means the debug interface was not claimed.
//! CMSIS-DAP v1 (HID) stays `unsupported`; this kind is only the interface walk.

const STLINK_VID: u16 = 0x0483;

pub(crate) fn open_failure_kind(vendor_id: u16, message: &str) -> &'static str {
    if vendor_id == STLINK_VID && stlink_interface_failure(message) {
        "stlink-interface"
    } else {
        "open-failed"
    }
}

fn stlink_interface_failure(message: &str) -> bool {
    let message = message.to_ascii_lowercase();
    message.contains("endpoint not found")
        || message.contains("endpointnotfound")
        || message.contains("interface not found")
        || message.contains("claiminterface")
        || message.contains("could not be claimed")
}

#[cfg(test)]
mod test {
    use super::open_failure_kind;

    #[test]
    fn endpoint_not_found_on_stlink_is_stlink_interface() {
        let message = "The debug probe could not be created. (ProbeCouldNotBeCreated(ProbeSpecific(EndpointNotFound)))";
        assert_eq!(open_failure_kind(0x0483, message), "stlink-interface");
    }

    #[test]
    fn hid_copy_is_not_reused_for_other_vendors() {
        assert_eq!(
            open_failure_kind(0x0d28, "USB endpoint not found"),
            "open-failed"
        );
    }

    #[test]
    fn unrelated_stlink_open_failure_stays_open_failed() {
        assert_eq!(
            open_failure_kind(0x0483, "permission denied"),
            "open-failed"
        );
    }
}
